### DISCUSSION TITLE
MonacoOutlineContribution extension

### DISCUSSION THREAD

**Discussion by BarryDrez** [eclipse-theia/theia#10652] [Q&A] [answered]
Hi,
I have created an extension of MonacoOutlineContribution since I want to customize the look of the Outline View when editing Java classes.  The problem is that it is being instantiated twice.  In my frontend module, I have the following:

```
    bind(MyOutlineContribution).toSelf().inSingletonScope();
    bind(FrontendApplicationContribution).toService(MyOutlineContribution);
    rebind(MonacoOutlineContribution).to(MyOutlineContribution).inSingletonScope();
```

Any ideas what I'm doing wrong?
Thanks in advance

**Comment by vince-fugnitto** [comment]
@BarryDrez thank you for the discussion, I believe I experienced the same behavior you had with your bindings. The fix would be not to bind the `FrontendApplicationContribution`.

Here is a small example:

```ts
import { FrontendApplication } from '@theia/core/lib/browser';
import { injectable } from '@theia/core/shared/inversify';
import { MonacoOutlineContribution } from '@theia/monaco/lib/browser/monaco-outline-contribution';

@injectable()
export class TheiaMonacoOutlineContribution extends MonacoOutlineContribution {

    onStart(app: FrontendApplication): void {
        super.onStart(app);
        console.log('custom theia monaco outline contribution started...');
    }
}
```

```ts
import { ContainerModule } from '@theia/core/shared/inversify';
import { MonacoOutlineContribution } from '@theia/monaco/lib/browser/monaco-outline-contribution';
import { TheiaMonacoOutlineContribution } from './theia-outline-contribution';

export default new ContainerModule((bind, unbind, isBound, rebind) => {
    bind(TheiaMonacoOutlineContribution).toSelf().inSingletonScope();
    rebind(MonacoOutlineContribution).to(TheiaMonacoOutlineContribution);
});
```

The following will only long once on `onStart` for me:

```
root INFO custom theia monaco outline contribution started...
```

If I add the `FrontendApplicationContribution` binding, I would get the two logs (which is what you experienced):

```ts
root INFO custom theia monaco outline contribution started...
root INFO custom theia monaco outline contribution started...
```

--

The reason is is because the contribution is already bound in the module by the default implementation, the goal in your case is to rebind `MonacoOutlineContribution` to  `MyOutlineContribution` only.

**Reply by BarryDrez** [reply]
Thanks for the reply @vince-fugnitto.  I tried as you suggest, and I can see that the onStart function is being called once; however, I am still seeing that the constructor is being called twice.  I have a constructor:
constructor() {
   console.log('custom theia monaco outline contribution constructed...');
}
in my custom outline contribution, and I see that it is called 2 times.  
The reason this is a problem for me is because I inject MyOutlineContribution into another class, and when I call a function of MyOutlineContribution  from the other class, the instance variables are all wrong.

regards - Barry

**Comment by vhemery** [comment, marked as answer]
Looks like a DI incorrect usage to me...

```
    // there, you are declaring MyOutlineContribution as a singleton service implemented by itself
    bind(MyOutlineContribution).toSelf().inSingletonScope();
    // you are telling FrontendApplicationContribution is implemented by the singleton service above
    bind(FrontendApplicationContribution).toService(MyOutlineContribution);
    // you are telling MonacoOutlineContribution is also implemented by the MyOutlineContribution class,
    // but not using the singleton service
    rebind(MonacoOutlineContribution).to(MyOutlineContribution).inSingletonScope();
```
As a consequence, when asking for injection of MonacoOutlineContribution, a new MyOutlineContribution is constructed instead of using the singleton.
My guess is that you would have probably wanted something like
```
    bind(MyOutlineContribution).toSelf().inSingletonScope();
    rebind(MonacoOutlineContribution).toService(MyOutlineContribution);
```
as for whether you should also bind it from FrontendApplicationContribution, I can't tell without looking at the actual MyOutlineContribution implementation. But if you have to, the binding was correctly written.

**Reply by BarryDrez** [reply]
Thank you @vhemery!  That worked.  Now there is only one instance.
